### PR TITLE
8320532: Remove Thread/ThreadGroup suspend/resume

### PR DESCRIPTION
**Pull request by AlanBateman**
The deadlock prone Thread/ThreadGroup suspend/resume were deprecated since JDK 1.2, deprecated for removal in Java 14, and re-specified/degraded to throw UnsupportedOperationException unconditionally in Java 19/20.  Early in Java 23 seems a fine time to finally remove these methods.

Corpus analysis of 176 million classes in 485k artifacts found no remaining usages of ThreadGroup.suspend/resume beyond the artifacts that include a copy of java.lang.ThreadGroup (!). It found 87 remaining uses of Thread.suspend and 86 remaining usages of Thread.resume, some of these are tests.

Thread.suspend/resume have always linked to the "Java Thread Primitive Deprecation" page. This originally explained the reasons why suspend/resume were deprecated. When these methods were degraded to throw UOE we changed the text to explain why the ability to suspend or resume a thread was removed. Now we must change it again. One choice is to re-word to explain why the Java APIs were removed or why the Java APIs don't define a way to suspend/resume threads, the other choice (which I prefer) is to remove the text.

The method description of java.lang.management.ThreadInfo.isSuspended is tweaked to link to JVMTI SuspendThread instead of Thread.suspend

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8320598](https://bugs.openjdk.org/browse/JDK-8320598) to be approved

### Issues
 * [JDK-8320532](https://bugs.openjdk.org/browse/JDK-8320532): Remove Thread/ThreadGroup suspend/resume (**Enhancement** - P4)
 * [JDK-8320598](https://bugs.openjdk.org/browse/JDK-8320598): Remove Thread/ThreadGroup suspend/resume (**CSR**)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [cc17ea42](https://git.openjdk.org/jdk/pull/16789/files/cc17ea42f8abc6c594be518508a83b531beb513f)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [cc17ea42](https://git.openjdk.org/jdk/pull/16789/files/cc17ea42f8abc6c594be518508a83b531beb513f)
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**) ⚠️ Review applies to [cc17ea42](https://git.openjdk.org/jdk/pull/16789/files/cc17ea42f8abc6c594be518508a83b531beb513f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16789/head:pull/16789` \
`$ git checkout pull/16789`

Update a local copy of the PR: \
`$ git checkout pull/16789` \
`$ git pull https://git.openjdk.org/jdk.git pull/16789/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16789`

View PR using the GUI difftool: \
`$ git pr show -t 16789`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16789.diff">https://git.openjdk.org/jdk/pull/16789.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16789#issuecomment-1825467760)

### CODE REVIEW

File: test/jdk/java/lang/Thread/ThreadStopTest.java [hunk 1]
```diff
@@ -22,60 +22,43 @@
  */
 
 /* @test
- * @bug  8289610 8249627 8205132
- * @summary Test that Thread stop/suspend/resume throw UOE
- * @run junit DegradedMethodsThrowUOE
```
Comment (dholmes-ora) [inline test/jdk/java/lang/Thread/ThreadStopTest.java:27]: You didn't really need to rename the test even though there is now only one degraded method left.

File: test/jdk/java/nio/channels/SocketChannel/SendUrgentData.java [hunk 2]
```diff
@@ -64,7 +64,9 @@ public static void main(String[] args) throws Exception {
         boolean inline = false;
         if (args.length > 0 && args[0].equals("-server")) {
             System.out.println(serverThread.getAddress());
-            Thread.currentThread().suspend();
```
Comment (dholmes-ora) [inline test/jdk/java/nio/channels/SocketChannel/SendUrgentData.java:67]: If suspend throws then surely this test can't be functioning correctly! ???

**Review by dholmes-ora — APPROVED** [review]
Looks good to me. I'm okay with removing the suspend/resume text from the deprecation page - no point trying to explain a 25+ year history.

Couple of queries on the test changes, but still approved.

Thanks

File: test/jdk/java/lang/Thread/ThreadStopTest.java (hunk 1 above)
Comment (AlanBateman) [reply test/jdk/java/lang/Thread/ThreadStopTest.java:27]: > You didn't really need to rename the test even though there is now only one degraded method left.

True but it's wasn't good name for a test that only exercises one method.

**Comment by AlanBateman** [comment]
> Looks good to me. I'm okay with removing the suspend/resume text from the deprecation page - no point trying to explain a 25+ year history.

Thanks, would you mind reviewing the CSR too? (The fixVersion is set to 23 so the skara bots aren't going to link the issues until we after the fork and the sync up of this branch).

**Review by jaikiran — APPROVED** [review]
The changes look fine to me. `SendUrgentData` is missing a copyright year update.

**Comment by sspitsyn** [comment]
Looks okay in general. Will make one more pass tomorrow.

**Review by sspitsyn — APPROVED** [review]

**Comment by stuart-marks** [comment]
Yeah, removing the text from the deprecation is reasonable. I expect that eventually the deprecation page will simply vanish when this is all over (I guess, after Thread.stop is finally removed).

Eventually somebody should write the definitive history of the Thread API.

**Review by stuart-marks — APPROVED** [review]

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**